File: Android_Firebase_Demo/app/src/main/java/firebase_demo/fg/com/android_firebase_demo/FirebaseEventListener.java
package firebase_demo.fg.com.android_firebase_demo;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;

public class FirebaseEventListener implements ChildEventListener {

    private DataListAdapter userList;


    public FirebaseEventListener(DataListAdapter userList) {
        this.userList = userList;
    }

    @Override
    public void onChildAdded(DataSnapshot dataSnapshot, String previousChildName) {
        User newUser = new User((String) dataSnapshot.getValue(), (String) dataSnapshot.getKey());
        if (!(userList.getUserList().contains(newUser))) {
            userList.addUser(newUser);
            userList.notifyDataSetChanged();
        }
    }

    @Override
    public void onChildMoved(DataSnapshot dataSnapshot, String s) {
        // Child moved
        // ...
    }

    @Override
    public void onChildRemoved(DataSnapshot dataSnapshot) {
        userList.removeUser(dataSnapshot.getKey());
        userList.notifyDataSetChanged();
    }

    @Override
    public void onChildChanged(DataSnapshot dataSnapshot, String s) {
        // Child changed
        // ...
    }

    @Override
    public void onCancelled(DatabaseError databaseError) {
        // Error
        // ...
    }
}

File: Android/app/src/main/java/com/example/floriangbh/firebase/FirebaseEventListener.kt
package com.example.floriangbh.firebase

import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.ChildEventListener


class FirebaseEventListener(private val itemListAdapter: ItemAdapter) : ChildEventListener {

    override fun onChildAdded(dataSnapshot: DataSnapshot, previousChildName: String?) {
        val newItem = ItemData(dataSnapshot.value as String? ?: "" , dataSnapshot.key ?: "")
        if (!itemListAdapter.partItemList.contains(newItem)) {
            itemListAdapter.addItem(newItem)
            itemListAdapter.notifyDataSetChanged()
        }
    }

    override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {
        // Child moved
        // ...
    }

    override fun onChildRemoved(dataSnapshot: DataSnapshot) {
        itemListAdapter.removeItem(dataSnapshot.key ?: "")
        itemListAdapter.notifyDataSetChanged()
    }

    override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {
        // Child changed
        // ...
    }

    override fun onCancelled(databaseError: DatabaseError) {
        // Error
        // ...
    }
}